src/reader.ts

```diff
@@ -46,42 +49,38 @@ export default class Reader<T extends Response> {
     const rawAddress = ipUtil.parse(ipAddress);
     const nodeCount = this.metadata.nodeCount;
 
-    // When storing IPv4 addresses in an IPv6 tree, they are stored as-is, so they
-    // occupy the first 32-bits of the address space (from 0 to 2**32 - 1).
-    // Which means they're padded with zeros.
-    const ipStartBit =
-      this.metadata.ipVersion === 6 && rawAddress.length === 4 ? 128 - 32 : 0;
-
     // Binary search tree consists of certain (`nodeCount`) number of nodes. Tree
     // depth depends on the ip version, it's 32 for IPv4 and 128 for IPv6. Each
     // tree node has the same fixed length and usually 6-8 bytes. It consists
     // of two records, left and right:
     // |         node        |
     // | 0x000000 | 0x000000 |
     let bit;
-    let nodeNumber = ipStartBit;
-    let pointer;
+    let nodeNumber = 0;
     let offset;
 
-    for (let i = ipStartBit; i < this.metadata.treeDepth; i++) {
+    let ipStartBit = 0;
+    // When storing IPv4 addresses in an IPv6 tree, they are stored as-is, so they
+    // occupy the first 32-bits of the address space (from 0 to 2**32 - 1).
+    // Which means they're padded with zeros.
+    if (rawAddress.length === 4) {
+      ipStartBit = this.ipv4StartBit;
+      nodeNumber = this.ipv4StartNodeNumber;
+    }
+
+    // Record value can point to one of three things:
+    // 1. Another node in the tree (most common case)
+    // 2. Data section address with relevant information (less common case)
+    // 3. Point to the value of `nodeCount`, which means IP address is unknown
+    for (let i = ipStartBit; i < this.metadata.treeDepth && nodeNumber < nodeCount; i++) {
       bit = ipUtil.bitAt(rawAddress, i - ipStartBit);
       offset = nodeNumber * this.metadata.nodeByteSize;
 
-      pointer = bit ? this.walker.right(offset) : this.walker.left(offset);
-
-      // Record value can point to one of three things:
-      // 1. Another node in the tree (most common case)
-      if (pointer < nodeCount) {
-        nodeNumber = pointer;
-
-        // 2. Data section address with relevant information (less common case)
-      } else if (pointer > nodeCount) {
-        return pointer;
+      nodeNumber = bit ? this.walker.right(offset) : this.walker.left(offset);
```

Comment:
Wouldn't it have somewhat noticeable performance impact? I mean you're doing max depth scans where's previously it was possible to exit earlier?

I can be wrong since there's no `if .. else ..` checks in the loop, which might be better. Would you mind running a performance test on it? https://github.com/runk/node-maxmind-benchmark